### ISSUE TITLE
question about plugin shortcodes

### ISSUE THREAD

**Issue by Jimmi08** [e107inc/e107#3211]
Yes, again, sorry.

I am trying to override download shortcodes from other plugin.  This is just example, because it was first shortcode.

I tried this in e_shortcode.php in jm_download plugin:

```
require_once(e_PLUGIN."download/download_shortcodes.php");
class jm_download_shortcodes extends download_shortcodes
{
  public $override = true; 
// when set to true, existing core/plugin shortcodes matching methods below 
   will be overridden.  
 function sc_download_admin_edit()
{
 MY CODE
} 
}
```
My plugin works with mine shortcodes, core download shortcodes and overriden shortcode.
 
![image](https://user-images.githubusercontent.com/5429548/41774516-5bab7c7e-7620-11e8-897b-8342a6bb072f.png)
But core download plugin displays only shortcode I used.
![image](https://user-images.githubusercontent.com/5429548/41774559-7c73009e-7620-11e8-98d2-2a36cc52cda8.png)

I didn't understand, download_shortcodes extends e_shortcode...  really not my area, so probably because I required that class again. To solve this I would need to copy all download shortcodes to my class.

So I tried:
```
//require_once(e_PLUGIN."download/download_shortcodes.php");
class jm_download_shortcodes extends e_shortcode
```
and then (order is important) :
```
$sc = e107::getScBatch('jm_download',true);
 $sc = e107::getScBatch('download',true); 
```
This way everything works.

Is the second one correct?  

Thanks

PS. What happens when 2 plugins override the same core shortcode? What last one is applied? Last installed? 

Too soon.  It worked because there were no variables.

**Comment by rica-carv** [comment]
When you override a core or plugin shortcode from another plugin, if it's donne in the e_shortcode.php file, it will be available to all.
If it's donne on the shortcode file of the plugin, it will be available only in the plugin itself, or where that shortcode file is included.
I'm currently doing that on a few custom plugins i have (the collection ones... :wink: ), which share some common shortcodes....

Regarding your question, yes, the working solution you find is currently the only one available...

**Comment by Jimmi08** [comment]
I still don't understand why extending download shortcodes class removes can't be used, because there is help:
// when set to true, existing core/plugin shortcodes matching methods below 

matching method -  but it changed all method, not only matching

**Comment by rica-carv** [comment]
They can´t be used because, as far i know, the plugin shortcode method is used only inside the plugin shortcode file. And therefore the class is not general.
The only shortcode class from a particular plugin available sitewide is the only used in e_shortcode, because that file is automatically loaded by e107 core, but the shortcode files specific of a plugin aren't.
Do i make myself clear or just added more confusion?

**Comment by Jimmi08** [comment]
No :)  
I used  class that extended download shortcodes class.  
I don't want to change anything global, I just don't want to copy already existed shortcodes in my plugin.  Download plugin can displays it's shortcodes.  But if I do this, download plugin doesn't display anything (no values in shortcodes).  - so it looks it overriden all shortcodes although it shouldn't do it. 

Or otherwise:
In custom plugin I can use my shortcodes and by extend class I can use other method (shortcodes) too. Yes, I need to have vars for them in my plugin.   But something like this influence that original plugin? It use it's own shortcode class.

**Comment by rica-carv** [comment]
Yes, you can use another plugin shortcodes in your plugin, but you need ALWAYS to include the original plugin shortcodes, and taht means include ALL the plugin shortcodes, not just one in particular.
That include can ONLY be anchieved via the `e107::getScBatch`, or use the php include, but it has some drawbacks, as you might already have noticed.
If all the logic vars for the shortcode is done from the main plugin file, you'll just have a bunch of shortcodes without the vars, wich you'll have to supply yourself from your own plugin.
Otherwise, in a perfect world, you can / should always perform that logic inside the shortcode php file, as it can be done without any trouble....

**Comment by myovchev** [comment]
It's mostly working as expected. There are two ways of SC usage:
- global (e_shortcode, any batch, everything registered globally)
- local - if the parsing piece of code is using the shortcodes as "extraCodes", you can look at `e_parse_shortcode#parseCodes()` to get the idea.

Additionally, if the parsing piece of code says `false` to parsing the global shortcodes while calling parse handler, only the local ones are used.
And it gets even more trickier - if the core routine for loading batch shortcodes is used, the retrieved batch object is registered and becomes global. 

I'm not aware of a way to override shortcodes with extend and `override = true` property nor any legit way plugin to override core class (@CaMer0n right?). What works for you is simply:
1. You load/register your shortcodes batch, the `download_admin_edit` shortcodes is registered globally
2. You load/register ALL core download shortcodes and because of the fact `download_admin_edit` already exists it's not registered and reference to your shortcode class is kept.

To summarize - the only reliable way for you to override global shortcode batch is via `e107_core/override` folder (and this can't be done by a plugin). In this case, you may extend the original shortcode class and override only the method(s) you want.

**Comment by Jimmi08** [comment]
Thank you for your answer.  
 
I wasn't probably clear.  I am not trying to override core shortcodes in download plugin itself. 

I want to use in my plugin my shortcodes plus already existing shortcodes (they work with the same data, $this->var).  So I used to extend this class and override some methods. It works with shorcodes in my plugin. But it removed data for download plugin itself.  It was weird, so I asked.  I tried to avoid copying the same code.  

When I have one template where I want to use shortcodes from more plugins, I can't (or I don't know how) to join them in one $sc and you can parse template only with one set of shortcodes. Parsing removes all {} without checking if this shortcode or not.

$override is here some years now and it works very nice for theme_shorcodes that are able to override core shortcodes.

**Comment by CaMer0n** [comment]
@myovchev As @Jimmi08 mentioned, theme_shortcodes can now override core shortcodes when `override = true`. There's an example with the news_summary shortcode (commented out) in e107_themes/bootstrap3/theme_shortcodes.php

**Comment by Jimmi08** [comment]
I don't need this anymore, so I am closing this.
